pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go

```diff
@@ -541,17 +554,52 @@ func (lbaas *LbaasV2) UpdateLoadBalancer(clusterName string, service *api.Servic
 		return err
 	}
 
-	// Anything left in addrs is a new member that needs to be added to a pool
-	for addr := range addrs {
-		_, err := v2_pools.CreateAssociateMember(lbaas.network, poolID, v2_pools.MemberCreateOpts{
-			Address:      addr,
-			ProtocolPort: int(ports[0].NodePort),
-			SubnetID:     lbaas.opts.SubnetId,
-		}).Extract()
-		if err != nil {
-			return err
+	waitLoadbalancerActiveProvisioningStatus(lbaas.network, loadbalancer.ID)
+	for _, port := range ports {
+		var listenerID, poolID string
+
+		// Find appropriate listener for this port
+		for _, listener := range listenersList {
+			if listener.ProtocolPort == int(port.Port) && listener.Protocol == string(port.Protocol) {
+				listenerID = listener.ID
+				break
+			}
+		}
+		if listenerID == "" {
+			return fmt.Errorf("Loadbalancer %s does not contain required listener for port %d and protocol %s", loadBalancerName, port.Port, port.Protocol)
+		}
+
+		// Find appropriate pool for found listener
+		for _, pool := range poolsList {
+			for _, currentListener := range pool.Listeners {
+				if listenerID == currentListener.ID {
+					poolID = pool.ID
+					break
+				}
+			}
+			if poolID != "" {
+				break
+			}
+		}
+		if poolID == "" {
+			return fmt.Errorf("Loadbalancer %s does not contain required pool for listener %s", loadBalancerName, listenerID)
+		}
+
+		for addr := range addrs {
```

Comment:
My understanding is that the UpdateLoadBalancer will ensure the set of members (hosts) for the existing load balancer are up to date.  Since there is one listener and one pool per load balancing port, this will not change in update since it is not intended to once a service has been created.  

The way this is implemented, the update will work as intended.  It is divided into two parts, the first removing members that are not required anymore and the second part adding new members in the correct listener/pool set.

I will see if this can be optimized some more.